torch_geometric/nn/models/lightgcn.py

```diff
@@ -0,0 +1,315 @@
+
+        nodes_fst = torch.index_select(out, 0, edge_label_index[0, :].long())
+        nodes_sec = torch.index_select(out, 0, edge_label_index[1, :].long())
+        out = torch.sum(nodes_fst * nodes_sec, dim=-1)
+
+        if self.objective == LightGCN.objectives[1]:
+            out = torch.sigmoid(out)
+
+        return out
+
+    def predict_link(self, edge_index: Adj, edge_label_index: Adj = None,
+                     prob: bool = False) -> Tensor:
+        """Predict links between nodes specified in :obj:`edge_label_index`.
+
+        This prediction head works only for the "link_prediction" objective.
+
+        Args:
+            edge_index (Adj): Edge tensor specifying the connectivity of the
+                graph.
+            edge_label_index (Adj, optional): Edge tensor specifying the start
+                and end nodes for which to compute rankings or proabilities.
+                If :obj:`edge_label_index` is set to :obj:`None`, all edges in
+                :obj:`edge_index` will be used (default: :obj:`None`).
+            prob (bool): Whether probabilities should be returned.
+        """
+        assert (self.objective == LightGCN.objectives[1]), 'Prediction ' \
+            'only works for "link_prediction" objective.'
+
+        if edge_label_index is None:
+            edge_label_index = edge_index
+
+        predictions = self.forward(edge_index, edge_label_index)
+        if prob:
+            return predictions
+
+        return torch.round(predictions)
+
+    def recommend(self, edge_index: Adj, source_indices: LongTensor,
+                  target_indices: LongTensor, topK: int = 1) -> Tensor:
+        """Get :obj:`topK` recommendations for nodes in :obj:`source_indices`.
+
+        This prediction head works only for the "ranking" objective.
+
+        Args:
+            edge_index (Adj): Edge tensor specifying the connectivity of the
+                graph.
+            source_indices (LongTensor): Nodes for which recommendations
+                should be generated (indices).
+            target_indices (LongTensor): Nodes which represent the choices
+                (indices).
+            topK (int, optional): Number of top recommendations to return
+                (default: 1).
+        """
+        assert (self.objective == LightGCN.objectives[0]), 'Recommendations ' \
+            'only work for "ranking" objective.'
+
+        prop_embeddings = self._propagate_embeddings(edge_index)
+        source_embeddings = prop_embeddings[source_indices]
+        target_embeddings = prop_embeddings[target_indices]
+
+        rankings = torch.mm(source_embeddings, target_embeddings.t())
+        top_indices = torch.topk(rankings, topK, dim=-1).indices
+        index_matrix = target_indices.reshape((1, -1))
+        index_matrix = index_matrix.repeat(source_indices.size(0), 1)
+
+        return index_matrix.gather(-1, top_indices)
+
+    def loss(self, preds: Tensor = None, edge_label: Tensor = None,
```

Comment:
Let's outsource that into `self.recommendation_loss` and `self.link_prediction_loss`.